### DISCUSSION TITLE
Why do I need to set QueryClient instance within React body for SSR?

### DISCUSSION THREAD

**Discussion by iicdii** [TanStack/query#4920] [Q&A] [answered]
https://tanstack.com/query/v4/docs/react/guides/ssr#using-hydration

It says

> Create a new QueryClient instance inside of your app, and on an instance ref (or in React state). **This ensures that data is not shared between different users and requests, while still only creating the QueryClient once per component lifecycle.**

And the example is

```typescript
// _app.jsx
import {
  Hydrate,
  QueryClient,
  QueryClientProvider,
} from '@tanstack/react-query'

export default function MyApp({ Component, pageProps }) {
  const [queryClient] = React.useState(() => new QueryClient())

  return (
    <QueryClientProvider client={queryClient}>
      <Hydrate state={pageProps.dehydratedState}>
        <Component {...pageProps} />
      </Hydrate>
    </QueryClientProvider>
  )
}
```

But what if I wrote like this, injecting the `queryClient` instance out of React scope? Does it make the data is shared between different requests and users? If so, how does it work?

```typescript
// _app.jsx
import {
  Hydrate,
  QueryClient,
  QueryClientProvider,
} from '@tanstack/react-query'

const queryClient = new QueryClient()

export default function MyApp({ Component, pageProps }) {
  return (
    <QueryClientProvider client={queryClient}>
      <Hydrate state={pageProps.dehydratedState}>
        <Component {...pageProps} />
      </Hydrate>
    </QueryClientProvider>
  )
}
```

**Comment by louis-young** [comment, marked as answer]
Hi 👋 

If you create the QueryClient instance in the top level module scope, it can share data between users as modules are evaluated once, meaning the QueryClient instance can persist across pre-renders on the server. If you create the QueryClient instance in function component scope then the instance can't be accidentally shared between users as each request/pre-render will have its own instance.

**Reply by iicdii** [reply]
Thanks!
I also add the conversation with ChatGPT for more detailed answer, because I found it interesting and useful. Correct me if the answer is wrong.

**Me**: What are the situations in which data can be shared between different requests and users mentioned in the React Query document?
**ChatGPT**:
> The data can be shared between different requests and users when the same instance of the QueryClient is used across multiple requests. **For example, if you create the instance of QueryClient outside of your main React component, such as at the top-level of your application or in a separate module, and import and use it in multiple components, it will be shared across all of those components, and therefore between multiple requests and users.**
This can lead to unexpected results, as different requests and users may interfere with each other's data and state, especially if they are making conflicting requests or modifying shared data. To avoid this, it's recommended to create a new instance of QueryClient for each request or user, so that their data and state are isolated and not shared between each other.
